src/tribol/mesh/MeshData.cpp

```diff
@@ -399,17 +405,31 @@ bool MeshData::computeFaceData( ExecutionMode exec_mode )
                   n[1][i] *= inv_mag;
 
                 } else if ( dim == 3 ) {
-                  // this method of computing an outward unit normal breaks the
-                  // face into triangular pallets by connecting two consecutive
-                  // nodes with the approximate centroid.
-                  // The average outward unit normal for the face is the average of
-                  // those of the pallets. This is exact for non-warped faces. To
-                  // compute the pallet normal, you only need edge vectors for the
-                  // pallet. These are constructed from the face centroid and the face
-                  // edge's first node and the face edge's two nodes
-
-                  // loop over num_nodes_per_elem-1 element edges and compute pallet
-                  // normal
+
+#ifdef TRIBOL_USE_ENZYME
+                  RealT x1[12];
+                  for ( int j = 0; j < num_nodes_per_elem; ++j ) {
+                    auto node_id = conn( i, j );
+                    for ( IndexT d{ 0 }; d < dim; ++d ) {
+                      x1[d * num_nodes_per_elem + j] = x[d][node_id];
+                    }
+                  }  // end loop over nodes
+
+                  // get vector n (normal of elem1)
+                  // NOTE: this limits this routine to quads
+                  RealT de1[3] = { -0.25 * x1[0] + 0.25 * x1[1] + 0.25 * x1[2] - 0.25 * x1[3],
+                                   -0.25 * x1[4] + 0.25 * x1[5] + 0.25 * x1[6] - 0.25 * x1[7],
+                                   -0.25 * x1[8] + 0.25 * x1[9] + 0.25 * x1[10] - 0.25 * x1[11] };
+                  RealT de2[3] = { -0.25 * x1[0] - 0.25 * x1[1] + 0.25 * x1[2] + 0.25 * x1[3],
+                                   -0.25 * x1[4] - 0.25 * x1[5] + 0.25 * x1[6] + 0.25 * x1[7],
+                                   -0.25 * x1[8] - 0.25 * x1[9] + 0.25 * x1[10] + 0.25 * x1[11] };
+                  n[0][i] = de1[1] * de2[2] - de1[2] * de2[1];
+                  n[1][i] = de1[2] * de2[0] - de1[0] * de2[2];
+                  n[2][i] = de1[0] * de2[1] - de1[1] * de2[0];
+                  RealT n_mag = std::sqrt( n[0][i] * n[0][i] + n[1][i] * n[1][i] + n[2][i] * n[2][i] );
```

Comment:
Should this be from the reference coords?

Reply:
Element normal is in terms of current coords; nodal normal magnitude is in terms of reference coords.